### ISSUE TITLE
Things with the new Program class

### ISSUE THREAD

**Issue by sophacles** [pyinvoke/invoke#276]
So there are a few things I've encountered with the new `Program` class. A few of them are bugs. A couple of them are usage things.

**Bugs:**

For the following bugs, please refer to this repository:  https://github.com/sophacles/inv-tests in the `inv_0.12 ` directory.

Quick overview: I create a module with a few library files, and main.py that defines a couple different program objects to serve as entry points into my task library. This module is installed into a virtual environment, along with the latest `invoke` from master. 

This is a reduction of my current invoke use case - a library of various task items and "ops-like" tasks, that are wrapped up into a few command line tools.

Here's the output I'm getting from this:

```
eheine@wirelessprvnat-172-17-31-229 ~/src/projects/invoke $ mkvirtualenv invtestNew python executable in invtest/bin/python2.7
# ...
Successfully installed invoke
Cleaning up...
(invtest)eheine@wirelessprvnat-172-17-31-229 ~/src/projects/invoke $ cd ../invthings/inv-tests/inv_0.12/lilprog 
(invtest)eheine@wirelessprvnat-172-17-31-229 ~/src/projects/inv-tests/inv_0.12/lilprog $ pip install -e .
# ---
Successfully installed lilprogtest
Cleaning up...
(invtest)eheine@wirelessprvnat-172-17-31-229 ~/src/projects/invthings/inv-tests/inv_0.12/lilprog $ myprog --list
No idea what '--list' is!
(invtest)eheine@wirelessprvnat-172-17-31-229 ~/src/projects/invthings/inv-tests/inv_0.12/lilprog $ otherprog --list
No idea what '--list' is!
(invtest)eheine@wirelessprvnat-172-17-31-229 ~/src/projects/invthings/inv-tests/inv_0.12/lilprog $ myprog subtask1
Traceback (most recent call last):
  File "/Users/eheine/envs/invtest/bin/myprog", line 9, in <module>
    load_entry_point('lilprogtest==0.1b1', 'console_scripts', 'myprog')()
  File "/Users/eheine/src/projects/invoke/invoke/program.py", line 246, in run
    self._parse(argv)
  File "/Users/eheine/src/projects/invoke/invoke/program.py", line 324, in _parse
    if self.args.list.value:
  File "/Users/eheine/src/projects/invoke/invoke/vendor/lexicon/__init__.py", line 23, in __getattr__
    return super(Lexicon, self).__getattr__(key)
  File "/Users/eheine/src/projects/invoke/invoke/vendor/lexicon/attribute_dict.py", line 7, in __getattr__
    raise AttributeError(key)
AttributeError: list
(invtest)eheine@wirelessprvnat-172-17-31-229 ~/src/projects/invthings/inv-tests/inv_0.12/lilprog $ myprog subtask2
Traceback (most recent call last):
  File "/Users/eheine/envs/invtest/bin/myprog", line 9, in <module>
    load_entry_point('lilprogtest==0.1b1', 'console_scripts', 'myprog')()
  File "/Users/eheine/src/projects/invoke/invoke/program.py", line 246, in run
    self._parse(argv)
  File "/Users/eheine/src/projects/invoke/invoke/program.py", line 324, in _parse
    if self.args.list.value:
  File "/Users/eheine/src/projects/invoke/invoke/vendor/lexicon/__init__.py", line 23, in __getattr__
    return super(Lexicon, self).__getattr__(key)
  File "/Users/eheine/src/projects/invoke/invoke/vendor/lexicon/attribute_dict.py", line 7, in __getattr__
    raise AttributeError(key)
AttributeError: list
(invtest)eheine@wirelessprvnat-172-17-31-229 ~/src/projects/invthings/inv-tests/inv_0.12/lilprog $ otherprog subtask1
Traceback (most recent call last):
  File "/Users/eheine/envs/invtest/bin/otherprog", line 9, in <module>
    load_entry_point('lilprogtest==0.1b1', 'console_scripts', 'otherprog')()
  File "/Users/eheine/src/projects/invoke/invoke/program.py", line 246, in run
    self._parse(argv)
  File "/Users/eheine/src/projects/invoke/invoke/program.py", line 324, in _parse
    if self.args.list.value:
  File "/Users/eheine/src/projects/invoke/invoke/vendor/lexicon/__init__.py", line 23, in __getattr__
    return super(Lexicon, self).__getattr__(key)
  File "/Users/eheine/src/projects/invoke/invoke/vendor/lexicon/attribute_dict.py", line 7, in __getattr__
    raise AttributeError(key)
AttributeError: list
(invtest)eheine@wirelessprvnat-172-17-31-229 ~/src/projects/invthings/inv-tests/inv_0.12/lilprog $ 

```

**Enhancements:**

* It would be nice to have the default `.yaml` (etc) file for a `Program` be `$BINARY.yaml` rather than `invoke.yaml`, or settable similar to `env_prefix`

* It would be nice to have a secondary "default" location, or a "location list" for secondary config files. For instance I'd like to point at a custom hierarcy (system config yaml for defaults, program config for overrides, homedir for further overriedes, etc - the main change being an overarching file for all the programs distributed with code that defines defaults across the set of tasks/programs).

* Core args access in tasks/context (this was discussed in #173 just bumping it here so it's not forgotten since that ticket is closed)

**Comment by bitprophet** [comment]
* Re: lack of `--list`, that's because Program strips out all of the `task_args` args when `namespace` is non-`None`, on the assumption that you don't want to be specifying "tasky" things.
    * **However** the same trigger causes `--help` to display the subcommands - see if that works for you.
    * tl;dr it behaves slightly less like core `invoke` but presumably in a manner more intuitive for people who don't know what invoke is.
        * Idea is "I want to distribute some Invoke tasks as 'a program' and my users shouldn't have to know much/anything about Invoke or its concepts in order to use it"
        * Obviously open to discussion about the merits of this approach...it's a big assumption
        * Probably easy to add an override switch in `Program.__init__` that says "I know I gave you a namespace but please act like `invoke` anyways and just use that namespace as the default instead of seeking `tasks.py`"...
* Re: it barfing trying to access `.list` in the main program logic, that is indeed a bug (stemming from the above). Presumably easy enough to fix, just have the same "has bundled namespace" test cause a skip of that particular block.
* Yup, I totally skipped over the configuration filename, presumably I would have run into that myself eventually w/ Fabric (so far I've only overridden the loaded task module name). Thanks for catching it.
* Can you clarify the "secondary default location", I don't grok it - the config stuff already has a hierarchy, are you asking for it to be changeable or what? An example might help.
* I've given Executor access to the core args already for the continuing work in #228, *and* tweaked how contexts/configs get generated, so adding that on shouldn't be too much more work. I'm still feeling out the "core `-H` flag ends up informing configgy things for Fabric hosts junk" which is very related.

**Comment by sophacles** [comment]
**list:**

re `--list`: Ok I see how to get it to work now. I think the conceptual decision is a good one going forward. I know I'll be re-enabling it because my users (including me) take advantage of `--list` regularly, and it's a part of muscle memory now - but I don't see that as a reason to clutter up `Program`. I can just re-enable it myself.

* Suggested refactoring: 
    * in main subclass `Program` as something like `InvokeProgram`
    * move `list_tasks()` to `InvokeProgram`
    * take advantage of `core_args()` to add `--list` `-e` etc
    * in `Program` create a method `handle_args()` containing all the `if args.X.value` logic/handling,
to call in `run` (between `_parse()` and `execute()`
     * in `InvokeProgram` override `handle_args` to include the `--list`(etc) things

Basically this would do a few things:

* Separate the `inv` program from `invoke` the toolkit
* Have `invoke` dogfood itself!
* Provide a nice example of how other tool writers can `Program` effectively
* keep the list of "preloaded core args" small

**other things**

* catching filename and executor - Awesome and thanks!
* "secondary default file location" -  I don't grok it anymore either. I broke my "don't type on the internet before coffee rule". That rule exists because of things like this bullet.


I'd be glad to take on the above refactoring if you like, just let me know.

**Comment by bitprophet** [comment]
Yea, I kept thinking along those lines re: the default functionality being the most stripped-down, then Invoke itself adding its own specific crap on top. Just never had the time. Thanks for digging into that! The PR is a big diff and I don't have time to dive into it before my vaca but I'll certainly poke it afterwards. :heart:

**Comment by kottenator** [comment]
I have the same problem (in `invoke==0.12.2`): I want to build my own program runner with pre-defined tasks. I create `script.py`:

```py
#!/bin/env python

import sys

from invoke import Program, __version__
from invoke.collection import Collection

import my_tasks

if __name__ == '__main__':
    sys.exit(Program(
        name="My Script",
        binary='script.py',
        version=__version__,
        namespace=Collection.from_module(my_tasks)
    ).run())
```

Then I run `script.py --help` and it works:

```
Usage: script.py [--core-opts] <subcommand> [--subcommand-opts] ...

Core options:

  --complete                     Print tab-completion candidates for given parse remainder.
  --hide=STRING                  Set default value of run()'s 'hide' kwarg.
  --write-pyc                    Enable creation of .pyc files.
  -d, --debug                    Enable debug output.
  -e, --echo                     Echo executed commands before running.
  -f STRING, --config=STRING     Runtime configuration file to use.
  -h [STRING], --help[=STRING]   Show core or per-task help and exit.
  -p, --pty                      Use a pty when executing shell commands.
  -V, --version                  Show version and exit.
  -w, --warn-only                Warn, instead of failing, when shell commands fail.

Subcommands:

  build
```

But when I run a particular command - **it fails**:

```sh
```

**Comment by darioblanco** [comment]
@kottenator I ran into the same issue, and I think the problem here is that when running your own program, you also need to specify some other default parameters within your arguments.

I extended then and it seems to be fixed
```python
class MyProgram(Program):
    def core_args(self):
        core_args = super(OrchestraProgram, self).core_args()
        print core_args
        extra_args = [
            Argument(names=['list'], help="Define list argument (does nothing)"),
            Argument(names=['no-dedupe'], help="Define no-dedupe argument (does nothing)"),
        ]
        return core_args + extra_args

program = MyProgram(name="myprogram", namespace=Collection.from_module(tasks), version=VERSION)
```

Is of course not ideal (same case when you have to set `Collection.from_module(tasks)` instead of `tasks` directly, and it works. I haven't digged more into how to extend `--list` and `--no-dedupe`, as I don't need to set those parameters, however, I think that invoke should offer them when creating your own app as well.

**Comment by bitprophet** [comment]
Linking this/#278 to #288 - when I get to solving the argument list split (core vs task related), it should be done within the context of these larger changes.
